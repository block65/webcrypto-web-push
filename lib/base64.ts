import {
  decode as decodeBase64,
  encode as encodeBase64,
} from 'base64-arraybuffer';

export { decodeBase64, encodeBase64 };

export function decodeBase64Url(str: string): ArrayBuffer {
  return decodeBase64(str.replace(/-/g, '+').replace(/_/g, '/'));
}

export function encodeBase64Url(arr: ArrayBuffer): string {
  return encodeBase64(arr)
    .replace(/\//g, '_')
    .replace(/\+/g, '-')
    .replace(/=+$/, '');
}

export function base64UrlToObject<T extends Record<string, unknown>>(
  str: string,
): T {
  return JSON.parse(new TextDecoder().decode(decodeBase64Url(str))) as T;
}

export function objectToBase64Url<T extends Record<string, unknown>>(
  obj: T,
): string {
  return encodeBase64Url(new TextEncoder().encode(JSON.stringify(obj)));
}
