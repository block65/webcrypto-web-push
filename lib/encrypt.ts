import { deriveClientKeys } from './client-keys.js';
import { hkdf } from './hkdf.js';
import { createInfo, createInfo2 } from './info.js';
import { ecJwkToBytes } from './jwk-to-bytes.js';
import { generateLocalKeys } from './local-keys.js';
import { getSalt } from './salt.js';
import type { PushSubscription } from './types.js';
import { arrayChunk, be16, flattenUint8Array, generateNonce } from './utils.js';

export interface EncryptedNotification {
  ciphertext: Uint8Array;
  salt: Uint8Array;
  serverPublic: Uint8Array;
}

// See https://developer.chrome.com/blog/web-push-encryption/
export async function encryptNotification(
  clientSubscription: PushSubscription,
  plaintext: Uint8Array,
): Promise<EncryptedNotification> {
  const clientKeys = await deriveClientKeys(clientSubscription);
  const salt = await getSalt();

  // Server keys
  const serverKeys = await generateLocalKeys();
  const serverPublic = ecJwkToBytes(serverKeys.publicJwk);

  const sharedSecret = await crypto.subtle.deriveBits(
    {
      name: 'ECDH',
      // @ts-expect-error - cloudflare worker encryption types are wrong?
      // namedCurve: 'P-256',
      public: clientKeys.publicKey,
    },
    serverKeys.privateKey,
    256,
  );

  // Infos
  const cekInfo = createInfo(clientKeys.publicBytes, serverPublic, 'aesgcm');
  const nonceInfo = createInfo(clientKeys.publicBytes, serverPublic, 'nonce');
  const keyInfo = createInfo2('auth');

  // Encrypt
  const ikmHkdf = await hkdf(clientKeys.authSecretBytes, sharedSecret);
  const ikm = await ikmHkdf.extract(keyInfo, 32);

  const messageHkdf = await hkdf(salt, ikm);
  const cekBytes = await messageHkdf.extract(cekInfo, 16);
  const nonceBytes = await messageHkdf.extract(nonceInfo, 12);

  const cekCryptoKey = await crypto.subtle.importKey(
    'raw',
    cekBytes,
    {
      name: 'AES-GCM',
      length: 128,
    },
    false,
    ['encrypt'],
  );

  const cipherChunks = await Promise.all(
    arrayChunk(plaintext, 4095).map(async (chunk, idx) => {
      const padSize = 0;
      const x = new Uint16Array([be16(padSize)]);
      const padded = new Uint8Array([
        ...new Uint8Array(x.buffer, x.byteOffset, x.byteLength),
        ...chunk,
      ]);

      const encrypted = await crypto.subtle.encrypt(
        {
          name: 'AES-GCM',
          iv: generateNonce(new Uint8Array(nonceBytes), idx),
        },
        cekCryptoKey,
        padded,
      );

      return new Uint8Array(encrypted);
    }),
  );

  return {
    ciphertext: flattenUint8Array(cipherChunks),
    salt,
    serverPublic,
  };
}
