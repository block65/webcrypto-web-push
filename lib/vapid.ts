import { sign, decodeBase64Url, encodeBase64Url } from './cf-jwt/main.js';
import { crypto } from './isomorphic-crypto.js';
import type { PushSubscription } from './types.js';

export interface VapidKeys {
  subject: string;
  publicKey: string;
  privateKey: string;
}

export async function vapidHeaders(
  subscription: PushSubscription,
  vapid: VapidKeys,
) {
  const vapidPublicKeyBytes = decodeBase64Url(vapid.publicKey);

  const publicKey = await crypto.subtle.importKey(
    'jwk',
    {
      kty: 'EC',
      crv: 'P-256',
      x: encodeBase64Url(vapidPublicKeyBytes.slice(1, 33)),
      y: encodeBase64Url(vapidPublicKeyBytes.slice(33, 65)),
      d: vapid.privateKey,
    },
    {
      name: 'ECDSA',
      namedCurve: 'P-256',
    },
    false,
    ['sign'],
  );

  const jwt = await sign(
    {
      aud: new URL(subscription.endpoint).origin,
      exp: Math.floor(Date.now() / 1000) + 12 * 60 * 60,
      sub: vapid.subject,
    },
    publicKey,
    {
      algorithm: 'ES256',
    },
  );

  return {
    headers: {
      authorization: `WebPush ${jwt}`,
      'crypto-key': `p256ecdsa=${vapid.publicKey}`,
    },
    // publicJwk,
  };
}
