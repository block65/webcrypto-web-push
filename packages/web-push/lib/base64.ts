import {
  base64ToUint8Array,
  stringToUint8Array,
  toUint8Array,
  uint8ArrayToBase64,
  uint8ArrayToString,
} from 'uint8array-extras';

export function toBase64UrlSafe(value: Uint8Array | ArrayBuffer) {
  return uint8ArrayToBase64(toUint8Array(value), { urlSafe: true });
}

export function base64ToObject<T extends Record<string, unknown>>(
  str: string,
): T {
  return JSON.parse(uint8ArrayToString(base64ToUint8Array(str)));
}

export function objectToBase64UrlSafe<T extends Record<string, unknown>>(
  obj: T,
): string {
  return toBase64UrlSafe(stringToUint8Array(JSON.stringify(obj)));
}
