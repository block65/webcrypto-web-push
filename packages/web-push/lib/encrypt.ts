import { deriveClientKeys } from './client-keys.js';
import { hkdf } from './hkdf.js';
import { createInfo, createInfo2 } from './info.js';
import { crypto } from './isomorphic-crypto.js';
import { ecJwkToBytes } from './jwk-to-bytes.js';
import { generateLocalKeys } from './local-keys.js';
import { getSalt } from './salt.js';
import type { PushSubscription } from './types.js';
import { arrayChunk, be16, flattenUint8Array, generateNonce } from './utils.js';

export interface EncryptedNotification {
  ciphertext: Uint8Array;
  salt: Uint8Array;
  localPublicKeyBytes: Uint8Array;
}

// See https://developer.chrome.com/blog/web-push-encryption/
export async function encryptNotification(
  subscription: PushSubscription,
  plaintext: Uint8Array,
): Promise<EncryptedNotification> {
  const clientKeys = await deriveClientKeys(subscription);
  const salt = await getSalt();

  // Local ephemeral keys
  const localKeys = await generateLocalKeys();
  const localPublicKeyBytes = ecJwkToBytes(localKeys.publicJwk);

  const sharedSecret = await crypto.subtle.deriveBits(
    {
      name: 'ECDH',
      // namedCurve: 'P-256',
      public: clientKeys.publicKey,
    },
    localKeys.privateKey,
    256,
  );

  // Infos
  const cekInfo = createInfo(
    clientKeys.publicKeyBytes,
    localPublicKeyBytes,
    'aesgcm',
  );
  const nonceInfo = createInfo(
    clientKeys.publicKeyBytes,
    localPublicKeyBytes,
    'nonce',
  );
  const keyInfo = createInfo2('auth');

  // Encrypt
  const ikmHkdf = await hkdf(clientKeys.authSecretBytes, sharedSecret);
  const ikm = await ikmHkdf.extract(keyInfo, 32);

  const messageHkdf = await hkdf(salt, ikm);
  const cekBytes = await messageHkdf.extract(cekInfo, 16);
  const nonceBytes = await messageHkdf.extract(nonceInfo, 12);

  const cekCryptoKey = await crypto.subtle.importKey(
    'raw',
    cekBytes,
    {
      name: 'AES-GCM',
      length: 128,
    },
    false,
    ['encrypt'],
  );

  const cipherChunks = await Promise.all(
    arrayChunk(plaintext, 4095).map(async (chunk, idx) => {
      const padSize = 0;
      const x = new Uint16Array([be16(padSize)]);
      const padded = new Uint8Array([
        ...new Uint8Array(x.buffer, x.byteOffset, x.byteLength),
        ...chunk,
      ]);

      const encrypted = await crypto.subtle.encrypt(
        {
          name: 'AES-GCM',
          iv: generateNonce(new Uint8Array(nonceBytes), idx),
        },
        cekCryptoKey,
        padded,
      );

      return new Uint8Array(encrypted);
    }),
  );

  return {
    ciphertext: flattenUint8Array(cipherChunks),
    salt,
    localPublicKeyBytes,
  };
}
