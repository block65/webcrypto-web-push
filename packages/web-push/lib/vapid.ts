import { base64ToUint8Array } from 'uint8array-extras';
import { crypto } from './isomorphic-crypto.js';
import type { PushSubscription } from './types.js';
import { invariant } from './utils.js';
import { toBase64UrlSafe } from './base64.js';
import { sign } from './jwt.js';

// undefined as its likely they are coming from env vars
// and this just makes DX nicer to check
export type VapidKeys = {
  subject: string | undefined;
  publicKey: string | undefined;
  privateKey: string | undefined;
};

export async function vapidHeaders(
  subscription: PushSubscription,
  vapid: VapidKeys,
) {
  invariant(vapid.subject, 'Vapid subject is empty');
  invariant(vapid.privateKey, 'Vapid private key is empty');
  invariant(vapid.publicKey, 'Vapid public key is empty');

  const vapidPublicKeyBytes = base64ToUint8Array(vapid.publicKey);

  const publicKey = await crypto.subtle.importKey(
    'jwk',
    {
      kty: 'EC',
      crv: 'P-256',
      x: toBase64UrlSafe(vapidPublicKeyBytes.slice(1, 33)),
      y: toBase64UrlSafe(vapidPublicKeyBytes.slice(33, 65)),
      d: vapid.privateKey,
    },
    {
      name: 'ECDSA',
      namedCurve: 'P-256',
    },
    false,
    ['sign'],
  );

  const jwt = await sign(
    {
      aud: new URL(subscription.endpoint).origin,
      exp: Math.floor(Date.now() / 1000) + 12 * 60 * 60,
      sub: vapid.subject,
    },
    publicKey,
    {
      algorithm: 'ES256',
    },
  );

  return {
    headers: {
      authorization: `WebPush ${jwt}`,
      'crypto-key': `p256ecdsa=${vapid.publicKey}`,
    },
    // publicJwk,
  };
}
