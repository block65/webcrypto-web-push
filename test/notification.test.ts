import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { decodeBase64Url, encodeBase64Url } from '../lib/base64.js';
import { encryptNotification } from '../lib/encrypt.js';
import { vapidHeaders } from '../lib/vapid.js';
import { fakeSubscriptions, fakeVapid } from './fixtures.js';

vi.mock('../lib/salt.js', () => ({
  getSalt: () => decodeBase64Url('4CQCKEyyOT_LysC17rsMXQ'),
}));

vi.mock('../lib/local-keys.js', () => ({
  generateLocalKeys: async () => {
    const publicJwk: JsonWebKey = {
      crv: 'P-256',
      ext: true,
      key_ops: ['deriveBits'],
      kty: 'EC',
      x: '7r5MIcg14B8yGrjhJAudH7c1mcB_B7af4KNwRJseU5w',
      y: 'Jkx_8JqbcwRJMd5zrZouJbIUe2AkFnhoDavSymNyOA0',
    };

    const privateJwk: JsonWebKey = {
      ...publicJwk,
      d: 'yFSH4q5h6mJ2VMryum5Nl31luqJQO8zzUJyIwgn0vVk',
    };

    return {
      publicJwk,
      privateKey: await crypto.subtle.importKey(
        'jwk',
        privateJwk,
        { name: 'ECDH', namedCurve: 'P-256' },
        true,
        ['deriveBits'],
      ),
      publicKey: await crypto.subtle.importKey(
        'jwk',
        publicJwk,
        { name: 'ECDH', namedCurve: 'P-256' },
        true,
        [],
      ),
    };
  },
}));

describe('', () => {
  beforeEach(() => {
    // tell vitest we use mocked time
    vi.useFakeTimers();
  });

  afterEach(() => {
    // restoring date after each test run
    vi.useRealTimers();
    vi.clearAllMocks();
  });

  test('encryptNotification', async () => {
    vi.setSystemTime(new Date(2000, 1, 1, 13));

    const subscription = fakeSubscriptions.test;

    const encryptResult = await encryptNotification(
      subscription,
      new TextEncoder().encode('This is test data.'),
    );

    const vapid = await vapidHeaders(subscription, fakeVapid);

    const requestInfo = {
      headers: {
        ...vapid.headers,

        'crypto-key': `keyid=p256dh;dh=${encodeBase64Url(
          encryptResult.serverPublic,
        )};${vapid.headers['crypto-key']}`,

        encryption: `keyid=p256dh;salt=${encodeBase64Url(encryptResult.salt)}`,

        'content-encoding': 'aesgcm',
        'content-length': encryptResult.ciphertext.byteLength.toString(),
        'content-type': 'application/octet-stream',
      },
      method: 'post',
      body: encryptResult.ciphertext,
    } satisfies RequestInit;

    expect(requestInfo.body).toMatchSnapshot();
    expect(requestInfo.method).toMatchSnapshot();

    const { authorization, ...headers } = requestInfo.headers;
    expect(headers).toMatchSnapshot();
    expect(authorization.match(/^(WebPush \w+\.\w+?\.)/)).toMatchSnapshot();
  });
});
