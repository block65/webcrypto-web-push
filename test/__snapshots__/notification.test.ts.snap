// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encryptNotification 1`] = `
Uint8Array [
  94,
  43,
  206,
  12,
  35,
  84,
  222,
  164,
  14,
  84,
  27,
  103,
  93,
  135,
  57,
  15,
  55,
  130,
  240,
  181,
  195,
  28,
  112,
  112,
  78,
  204,
  132,
  168,
  189,
  200,
  158,
  205,
  71,
  53,
  254,
  197,
]
`;

exports[`encryptNotification 2`] = `"post"`;

exports[`encryptNotification 3`] = `
{
  "content-encoding": "aesgcm",
  "content-length": "36",
  "content-type": "application/octet-stream",
  "crypto-key": "keyid=p256dh;dh=BO6-TCHINeAfMhq44SQLnR-3NZnAfwe2n-CjcESbHlOcJkx_8JqbcwRJMd5zrZouJbIUe2AkFnhoDavSymNyOA0;p256ecdsa=BKKXE3jJV5UJ6c8HVPam6DvMPGZK26r-M7ojsO2T_KdjdeMT2d7oQpaO-VI3o3wn33mQ8JlHta3OSJ5f67Ac5ZY",
  "encryption": "keyid=p256dh;salt=4CQCKEyyOT_LysC17rsMXQ",
}
`;

exports[`encryptNotification 4`] = `
[
  "WebPush eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJpYXQiOjk0OTM4MTIwMCwiYXVkIjoiaHR0cHM6Ly9mY20uZ29vZ2xlYXBpcy5jb20iLCJleHAiOjk0OTQyNDQwMCwic3ViIjoibWFpbHRvOnRlc3RAdGVzdC50ZXN0In0.",
  "WebPush eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJpYXQiOjk0OTM4MTIwMCwiYXVkIjoiaHR0cHM6Ly9mY20uZ29vZ2xlYXBpcy5jb20iLCJleHAiOjk0OTQyNDQwMCwic3ViIjoibWFpbHRvOnRlc3RAdGVzdC50ZXN0In0.",
]
`;
